### PR TITLE
[HWKMETRICS-564] fix several bugs

### PR DESCRIPTION
**Pull request by jsanda**
When jobs fall behind and are rescheduled for immediate execution, we first
renew the job lock. The job should only be executed again if the lock is
successfully renewed. We were always executing the job. This is fixed.

When there are jobs to execute across multiple time slices, we were executing
things out of order. This is fixed simply by replacing flatMap with concatMap
in a few places.

When retrying a job execution, we were updating finished_jobs_idx with the
wrong timestamp. We were using the trigger time instead of the time in which
the job is scheduled. In this scenario, the scheduled time != the trigger time.

### CODE REVIEW

File: job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java [hunk 1]
```diff
@@ -140,7 +140,8 @@
     static final int SCHEDULING_LOCK_TIMEOUT_IN_SEC = 5;
 
     // TODO We probably want this to be configurable
-    static final int JOB_EXECUTION_LOCK_TIMEOUT_IN_SEC = 3600;
+//    static final int JOB_EXECUTION_LOCK_TIMEOUT_IN_SEC = 3600;
```
Comment (stefannegrea) [inline job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java]: This was confusing change, but from what I understand needs to be reverted.

File: job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java [hunk 2]
```diff
@@ -269,27 +272,42 @@ public void start() {
...
                             .flatMap(this::acquireJobLock)
                             .filter(jobLock -> jobLock.acquired)
                             .map(jobLock -> jobLock.jobDetails)
-                            .doOnNext(details -> logger.debugf("Acquired job lock for %s in time slice %s", details,
-                                    timeSliceLock.timeSlice))
-                            .flatMap(details -> executeJob(details, timeSliceLock.timeSlice, activeJobs).toObservable()
+                            .doOnNext(details -> {
+                                logger.debugf("Acquired job lock for %s in time slice %s", details,
+                                        timeSliceLock.timeSlice);
+                                synchronized (activeJobsLock) {
```
Comment (stefannegrea) [inline job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java]: Why is this lock needed with a ConcurrentSkipListSet?

File: job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java (hunk 1 above)
Comment (jsanda) [inline job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java]: Good catch. Yes, it needs to be reverted. Right now it is hard coded, but that will change with [HWKMETRICS-551](https://issues.jboss.org/browse/HWKMETRICS-551).

File: job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java (hunk 2 above)
Comment (jsanda) [inline job-scheduler/src/main/java/org/hawkular/metrics/scheduler/impl/SchedulerImpl.java]: I don't think the lock is needed.

**Comment by stefannegrea** [comment]
Thank you @jsanda!

**Pull request merged**